sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala

```diff
@@ -83,15 +85,94 @@ trait FunctionRegistry {
 
   /** Clear all registered functions. */
   def clear(): Unit
+}
 
-  /** Create a copy of this registry with identical functions as this registry. */
-  override def clone(): FunctionRegistry = throw new CloneNotSupportedException()
+object FunctionRegistryBase {
+
+  /**
+   * Return an expression info and a function builder for the function as defined by
+   * T using the given name.
+   */
+  def build[T : ClassTag](name: String): (ExpressionInfo, Seq[Expression] => T) = {
```

Comment:
nit: what if we simply put it in the `trait FunctionRegistryBase[T]`?

Reply:
Please correct me if I am wrong here. Scala doesn't seem to allow calling the methods in a generic trait unless the class/object extends it and implements all abstract methods. Since we need to call these methods (build and expressionInfo) inside the companion objects for FunctionRegistry and TableFunctionRegistry, they can't be in the `FunctionRegistryBase[T]` trait.

Reply:
You are right, I missed this part.